src/fheroes2/ai/ai_hero_action.cpp

```diff
@@ -376,22 +376,49 @@ namespace
             DEBUG_LOG( DBG_AI, DBG_INFO, hero.GetName() << " attacks the enemy castle " << castle->GetName() )
 
             Heroes * defender = castle->GetHero();
-            const bool playVanishingHeroSound = defender != nullptr && defender->isControlHuman();
+            const bool isDefenderHuman = castle->isControlHuman();
+
+            if ( defender && ( defender->isControlHuman() != isDefenderHuman ) ) {
+                assert( 0 );
+
+                if ( isDefenderHuman ) {
+                    DEBUG_LOG( DBG_AI, DBG_INFO, defender->GetName() << " the AI hero is protecting the human castle " << castle->GetName() )
```

Comment:
If something happens that shouldn't normally happen, it's better to use `DBG_WARN`:

```suggestion
                    DEBUG_LOG( DBG_AI, DBG_WARN, defender->GetName() << " the AI hero is protecting the human castle " << castle->GetName() )
```